extensions/amp-story/1.0/amp-story-grid-layer.js

```diff
@@ -134,6 +153,29 @@ export class AmpStoryGridLayer extends AmpStoryBaseLayer {
     });
   }
 
+  /**
+   * Styles text with a background color based on the value of
+   * the text-background-color attribute
+   * @private
+   */
+  setDescendentCssTextStyles_() {
+    const elementsToUpgradeStyles = scopedQuerySelectorAll(this.element,
+        TEXT_BACKGROUND_COLOR_SELECTOR);
+
+    Array.prototype.forEach.call(elementsToUpgradeStyles, element => {
+      const color = element.getAttribute(TEXT_BACKGROUND_COLOR_ATTRIBUTE_NAME);
+      const attributes = dict({
+        'style': `background-color:${color}`,
+      });
+      const spanEl = createElementWithAttributes(
+          /** @type {!Document} */ (this.element.ownerDocument),
+          'span', attributes);
+      spanEl.textContent = element.textContent;
```

Comment:
This PR is overall awesome!

The only concern I have is about this very line, that could change the structure of the DOM.
For example if a publisher writes the following example, because only the `textContent` is copied, the inner `<span>` would get removed and only the text content would remain. In this example, the word "pink" would not be pink.
```
<style amp-custom>
.pink {
  color: pink;
}
</style>

// Publisher code
<amp-story-grid-layer>
  <p text-background-color="rgba(255, 255, 255, 0.6)">
    <span class="pink">Pink</span> sneakers are cool
  </p>
</amp-story-grid-layer>

// Transformed into
<amp-story-grid-layer>
  <p text-background-color="rgba(255, 255, 255, 0.6)">
    <span>
      Pink sneakers are cool
    </span>
  </p>
</amp-story-grid-layer>
```

One way to work around it would be to move all the `element`'s children into the `spanEl`, like we do here https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/1.0/amp-story-page-attachment.js#L152-L154


But doing so creates another issue where only the direct descendants get the `background-color` ([prototype here](https://codepen.io/gmajoulet/pen/YbqdLj)). I'm fine with this as long as we mention it in the documentation
```
<p text-background-color="rgba(...)">
  All the cool text.
  <p> <!-- This would not get styled -->
    And even more cool text.
  </p>
</p>
```

@estherkim @newmuis wdyt?

Reply:
(It's quite possible I've not thought this through) can we just set the style on the original `element` directly?

Reply:
@newmuis That's what I was going for at first - not changing the DOM would be ideal. But  `display: inline !important` doesn't work unless I changed the parent element from `display: grid` to `display: block`. I even tried using `text-shadow` but couldn't come up with a clean solution.

@gmajoulet I like the workaround you suggested of moving all the `element`'s children into the `spanEl`. I can still traverse down and style all the children if that's more intuitive to publishers (and skip the elements that already specify a background color). Also I like the rounded corners and space between lines from your prototype, I'll add that :)

Reply:
Good idea!
The only question is whether we want to support n levels of nested children, or only the direct descendants. I'd vote for the latter to keep the code easier and execution faster, and communicate this clearly in the documentation. If it ever becomes an issue for some publishers or creation tools, we can always improve this and support N level of descendants.

Reply:
I think we have to be careful with moving elements around -- some AMP components make assumptions about their parents/ancestors and don't like to be reparented 🙁  This is particularly problematic because we can't guarantee that `amp-story-grid-layer` will be built before those other elements, so they could be initialized with the wrong DOM tree hierarchy.

Maybe we can either (a) add validation rules to disallow certain types of content, or (b) add JS assertions with console errors?

Reply:
What if we put the onus on publishers to change the DOM themselves? That means we'd only be providing some styling with `[text-background-color]` but we could have examples on how to use it.

Reply:
Agree the onus should be on the publisher here!  Would just be good to make it either (a) work, or (b) fail reliably; the only case I'm worried about is race conditions that could come from the unspecified build order.

@gmajoulet had an idea on how to leverage a design pattern we used for `<amp-story-page-attachment>` that could make this work

Reply:
So! After brainstorming with @gmajoulet, we decided to simply provide the `[text-background-color]` for styling, and let the publishers use it however they'd like, without changing the DOM structure. I created an example story with different use cases and I'll update the MD file next, when I update the validator.

I don't really follow re: race conditions, @newmuis can you give an example?

Reply:
I believe the race conditions Jon mentioned can happen when (1) an AMP element is built, and (2) you reparent it. Because you moved away from the initial solution that was changing the DOM structure, you shouldn't have to worry about it now